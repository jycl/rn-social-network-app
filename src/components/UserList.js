import React, { Component } from "react";
import {
  View,
  FlatList,
  Text,
  StyleSheet,
  TouchableOpacity
} from "react-native";
import PropTypes from "prop-types";
import colors from "../styles/colors";
import fontSize from "../styles/fontSize";

/**
 * UserList component renders a FlatList with data from the userList props.
 * The list displays users' names retrieved from the backend and allows the
 * app user to press the list item to navigate to get more details (via onPress props).
 *
 * Props:
 * @property {array} userList array of objects that each represent a user and their details/props
 * @property {func} onPress callback method invoked when item row for user is pressed
 *
 * @author Joshua Leung <joshuaycleung@gmail.com>
 */
class UserList extends Component {
  render() {
    return (
      <FlatList
        ref="userList"
        keyExtractor={this._keyExtractor}
        renderItem={this._renderUserRow}
        ItemSeparatorComponent={this._renderSeparator}
        data={this.props.userList}
      />
    );
  }

  /**
   * Render each list item as button passing row item into props.onPress
   */
  _renderUserRow = ({ item, index }) => {
    return (
      <TouchableOpacity
        onPress={() => this.props.onPress(item)}
        style={styles.listItemContainer}
      >
        <Text style={styles.listItemText}>{item.name}</Text>
      </TouchableOpacity>
    );
  };

  /**
   *  Seperator component rendered between each row
   */
  _renderSeparator = () => <View style={styles.separator} />;

  /**
   *  Set list item key by the user Id from backend and its index
   */
  _keyExtractor = (item, index) => `${item.id}_${index}`;
}

const styles = StyleSheet.create({
  separator: {
    backgroundColor: colors.darkBlue,
    height: 2,
    width: "100%"
  },
  listItemContainer: {
    padding: 10,
    backgroundColor: colors.blue
  },
  listItemText: {
    color: colors.cream,
    paddingVertical: 10,
    fontSize: fontSize.medium
  }
});

UserList.propTypes = {
  userList: PropTypes.array,
  onPress: PropTypes.func
};

UserList.defaultProps = {
  userList: [],
  onPress: () => {}
};

export default UserList;
