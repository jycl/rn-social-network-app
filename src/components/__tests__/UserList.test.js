import React from "react";
import { FlatList } from "react-native";
import UserList from "../UserList";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

describe("Rendering user list with mock props", () => {
  beforeAll(() => {
    configure({ adapter: new Adapter() });
  });

  //test set of data
  const userList = [
    {
      id: 1,
      name: "John Doe"
    },
    {
      id: 2,
      name: "Kevin McCormick"
    }
  ];

  it("UserList renders correctly and updates data", () => {
    const component = shallow(<UserList />);
    expect(component).toMatchSnapshot();
    expect(component.find(FlatList).props().data).toHaveLength(0);
    //update user list and see if FlatList data updates

    component.setProps({ userList });
    expect(component).toMatchSnapshot();
    expect(component.find(FlatList).props().data).toHaveLength(2);
  });
});
