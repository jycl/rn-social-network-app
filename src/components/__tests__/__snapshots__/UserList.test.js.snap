// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering user list with mock props UserList renders correctly and updates data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserList
    onPress={[Function]}
    userList={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "ItemSeparatorComponent": [Function],
      "data": Array [],
      "disableVirtualization": false,
      "horizontal": false,
      "initialNumToRender": 10,
      "keyExtractor": [Function],
      "maxToRenderPerBatch": 10,
      "numColumns": 1,
      "onEndReachedThreshold": 2,
      "renderItem": [Function],
      "scrollEventThrottle": 50,
      "updateCellsBatchingPeriod": 50,
      "windowSize": 21,
    },
    "ref": "userList",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "ItemSeparatorComponent": [Function],
        "data": Array [],
        "disableVirtualization": false,
        "horizontal": false,
        "initialNumToRender": 10,
        "keyExtractor": [Function],
        "maxToRenderPerBatch": 10,
        "numColumns": 1,
        "onEndReachedThreshold": 2,
        "renderItem": [Function],
        "scrollEventThrottle": 50,
        "updateCellsBatchingPeriod": 50,
        "windowSize": 21,
      },
      "ref": "userList",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Rendering user list with mock props UserList renders correctly and updates data 2`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserList
    onPress={[Function]}
    userList={
      Array [
        Object {
          "id": 1,
          "name": "John Doe",
        },
        Object {
          "id": 2,
          "name": "Kevin McCormick",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "ItemSeparatorComponent": [Function],
      "data": Array [
        Object {
          "id": 1,
          "name": "John Doe",
        },
        Object {
          "id": 2,
          "name": "Kevin McCormick",
        },
      ],
      "disableVirtualization": false,
      "horizontal": false,
      "initialNumToRender": 10,
      "keyExtractor": [Function],
      "maxToRenderPerBatch": 10,
      "numColumns": 1,
      "onEndReachedThreshold": 2,
      "renderItem": [Function],
      "scrollEventThrottle": 50,
      "updateCellsBatchingPeriod": 50,
      "windowSize": 21,
    },
    "ref": "userList",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "ItemSeparatorComponent": [Function],
        "data": Array [
          Object {
            "id": 1,
            "name": "John Doe",
          },
          Object {
            "id": 2,
            "name": "Kevin McCormick",
          },
        ],
        "disableVirtualization": false,
        "horizontal": false,
        "initialNumToRender": 10,
        "keyExtractor": [Function],
        "maxToRenderPerBatch": 10,
        "numColumns": 1,
        "onEndReachedThreshold": 2,
        "renderItem": [Function],
        "scrollEventThrottle": 50,
        "updateCellsBatchingPeriod": 50,
        "windowSize": 21,
      },
      "ref": "userList",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
